spec/lib/ldap_group_reader_spec.rb

```diff
@@ -0,0 +1,92 @@
+require 'rails_helper'
+require 'ldap_group_reader'
+
+RSpec.describe LDAPGroupReader do
+  let(:ldap) { instance_double(Net::LDAP) }
+  let(:connection) { instance_double(Devise::LDAP::Connection, ldap: ldap) }
+  let(:group_base) { 'ou=group,dc=sanger,dc=ac,dc=uk' }
+  let(:person_base) { 'ou=people,dc=sanger,dc=ac,dc=uk' }
+  let(:member_attr) { 'member' }
+  let(:ldap_config) do
+    {
+      'base' => person_base,
+      'group_base' => group_base,
+      'group_membership_attribute' => member_attr,
+    }
+  end
+
+  before do
+    allow(Devise::LDAP::Adapter).to receive(:ldap_connect).and_return(connection)
+    allow(Rails.application.config).to receive(:ldap).and_return(ldap_config)
+  end
+
+  describe '#fetch_members' do
+    let(:group_name) { 'mygroup' }
+    let(:group_filter) {
+      Net::LDAP::Filter.eq('cn', group_name) & Net::LDAP::Filter.eq('objectclass', 'posixGroup')
+    }
+
+    context 'when the group does not exist' do
+      before do
+        expect(ldap).to receive(:search).
+          with(filter: group_filter, base: group_base, attributes: [member_attr]).
+          and_return([])
+      end
+
+      it 'should return an empty array' do
+        expect(LDAPGroupReader.fetch_members(group_name)).to be_empty
+      end
+    end
+
+    context 'when the group exists' do
+      let(:ldap_group) do
+        lg = double('ldap-group')
+        allow(lg).to receive(member_attr).and_return(ldap_group_members)
+        lg
+      end
+
+      before do
+        expect(ldap).to receive(:search).
+          with(filter: group_filter, base: group_base, attributes: [member_attr]).
+          and_return([ldap_group])
+      end
+
+      context 'when the group has no members' do
+        let(:ldap_group_members) { [] }
+
+        it 'should return an empty array' do
+          expect(LDAPGroupReader.fetch_members(group_name)).to be_empty
+        end
+      end
+
+
+      context 'when the group has members' do
+        let(:member_uids) { ['alpha', 'beta', 'gamma'] }
+        let(:ldap_group_members) { member_uids.map { |uid| "uid=#{uid},#{person_base}" } }
+
+        let(:ldap_people) do
+          member_uids.map { |uid| double('ldap-person', cn: ["#{uid} mc#{uid}"], mail: ["#{uid}@place.com"]) }
+        end
+
+        let(:person_filter) {
+          Net::LDAP::Filter.eq('uid', 'alpha') | Net::LDAP::Filter.eq('uid', 'beta') | Net::LDAP::Filter.eq('uid', 'gamma')
```

Comment:
Normally, the most common `or` operators that are in use in ruby are `or` or ` ||`.
The operator `|` is bitwise operator, it is equivalent to `|| `when operating with booleans, but differs when operating with other datatypes. For example, is not the same `3|4` than `3||4`. In this case `Net::LDAP::Filter.eq('uid', 'alpha')` is boolean, so it's not a mistake, but could give not expected results in other case.